samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -161,72 +127,102 @@ class KafkaSystemProducer(systemName: String,
-                sourceData.exceptionInCallback.compareAndSet(
-                  null,
-                  new SamzaException("Unable to send message from %s to system %s." format(source, systemName),
-                    exception))
-
-                metrics.sendFailed.inc
-                error("Unable to send message on Topic:%s Partition:%s" format(topicName, partitionKey),
-                             exception)
-              }
-            }
-          })
-        sourceData.latestFuture = futureRef
-      }
+      currentProducer.send(record, new Callback {
+        def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
+          if (exception == null) {
+            //send was successful.
+            metrics.sendSuccess.inc
+          }
+          else {
+            val producerException = new SystemProducerException(KafkaSystemProducer.FailedSendMsgFormat
+              .format(source, systemName, topicName, partitionKey), exception)
+
+            handleSendException(currentProducer, producerException, true)
+          }
+        }
+      })
       metrics.sends.inc
     } catch {
-      case e: Exception => {
-        error("Closing the producer because of an exception in send: ", e)
-
-        closeAndNullifyCurrentProducer(currentProducer)
+      case e : Exception =>
+        val producerException = new SystemProducerException(KafkaSystemProducer.FailedSendMsgFormat
+          .format(source, systemName, topicName, partitionKey), e)
 
-        metrics.sendFailed.inc
-        throw new SamzaException(("Failed to send message on Topic:%s Partition:%s Exception:\n %s,")
-          .format(topicName, partitionKey, e))
-      }
+        handleSendException(currentProducer, producerException, isFatalException(e))
+        throw producerException
     }
   }
 
+
   def flush(source: String) {
     updateTimer(metrics.flushNs) {
       metrics.flushes.inc
 
-      val sourceData = sources.get(source)
-      //if latestFuture is null, it probably means that there has been no calls to "send" messages
-      //Hence, nothing to do in flush
-      if(sourceData.latestFuture != null) {
-        while(!sourceData.latestFuture.isDone && sourceData.exceptionInCallback.get() == null) {
-          try {
-            sourceData.latestFuture.get()
-          } catch {
-            case t: Throwable => error(t.getMessage, t)
-          }
-        }
+      if (producer == null) {
+        throw new SystemProducerException(KafkaSystemProducer.NullProducerMsg)
+      }
 
-        //if there is an exception thrown from the previous callbacks just before flush, we have to fail the container
-        if (sourceData.exceptionInCallback.get() != null) {
-          metrics.flushFailed.inc
-          throw sourceData.exceptionInCallback.get()
-        } else {
-          trace("Flushed %s." format (source))
+      // Flush only throws InterruptedException, all other errors are handled in send() callbacks
+      producer.flush()
```

Comment:
1. currentProducer = producer to avoid NPE w/ concurrent stop.
2. Wondering if it's ok to flush a closed producer?

Reply:
2. Yep it's fine. I checked with the kafka team.